pkg/daemon/daemon.go

```diff
@@ -177,38 +226,116 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
 	return nil
 }
 
-// handleNodeUpdate is the handler for informer callbacks detecting node
-// changes. If an update is requested by the controller, we assume that
-// that means something changed and apply the desired config no matter what.
+// runOnce pulls the MachineConfig from either a file or remote URL
+// executes one run, and exits.
+// TODO: Revisit the Run/process methods and refactor and unify them with runOnce
+func (dn *Daemon) runOnce() error {
+	var machineConfig *mcfgv1.MachineConfig
+	var oldConfig mcfgv1.MachineConfig
+	var err error
+
+	if strings.HasPrefix("http://", dn.onceFrom) || strings.HasPrefix("https://", dn.onceFrom) {
+		// NOTE: This case expects a cluster to exists already.
+		// If we sense a remote URL has been provided then request MC content
+		// from a remote URL and parse it.
+		glog.V(2).Infof("Getting machine config content from %s", dn.onceFrom)
+		machineConfig, err = dn.getMachineConfigFromURL(dn.onceFrom)
+		if err != nil {
+			return err
+		}
+		needUpdate, err := dn.prepUpdateFromCluster()
+		if err != nil {
+			glog.V(2).Infof("Unable to prep update: %s", err)
+			return err
+		} else if needUpdate == false {
+			glog.V(2).Infof("No update needed")
+			return nil
+		}
+		// At this point we have verified we need to update
+		if err := dn.executeUpdateFromClusterWithMachineConfig(machineConfig); err != nil {
+			glog.Warningf("Unable to update: %s", err)
+			return err
+		}
+		return nil
+
+	} else if ValidPath(dn.onceFrom) {
+		// NOTE: This case expects that the cluster is NOT CREATED YET.
+		// If we sense a local file has been provided parse it.
+		oldConfig = mcfgv1.MachineConfig{}
+		absoluteOnceFrom, err := filepath.Abs(filepath.Clean(dn.onceFrom))
+		if err != nil {
+			return err
+		}
+		machineConfig, err = dn.getMachineConfigFromFile(absoluteOnceFrom)
+		if err != nil {
+			return err
+		}
+		// Execute update without hitting the cluster
+		return dn.update(&oldConfig, machineConfig)
+	}
+	// Otherwise return an error as the input format is unsupported
+	return fmt.Errorf("%s is not a path nor url; can not run once", dn.onceFrom)
+}
+
+// handleNodeUpdate is the gatekeeper handler for informer callbacks detecting
+// node changes. If an update is requested by the controller, we assume that
+// that means something changed and pass over to execution methods no matter what.
 // Also note that we only care about node updates, not creation or deletion.
 func (dn *Daemon) handleNodeUpdate(old, cur interface{}) {
 	node := cur.(*corev1.Node)
 
 	// First check if the node that was updated is this daemon's node
-	if (node.Name != dn.name) {
-		// The node that was changed was not ours
-		return
+	if node.Name == dn.name {
+		// Pass to the shared update prep method
+		_, err := dn.prepUpdateFromCluster()
+		if err != nil {
+			glog.V(2).Infof("Unable to prep update: %s", err)
+			return
```

Comment:
So for this and the following error, note that this is a callback function, so the return actual doesn't go anywhere useful. In that case, if say, the `prepUpdate` fails, we just return to... listening to the informer, which will trigger this again very soon, and it will just silently fail infinitely.

Barring a better way to do this, I recommend that on any update failure, for the time being to set the node to `degraded` and reboot the machine, so we can "catch" the error and sleep. I recognize this may not be the most clean way to handle this, and an alternative I tried was to throw/catch a `panic` and handle it that way if you think that makes sense.

Reply:
Hrm, that's a good point. `panic` handling is kind of a last ditch effort IMHO. I'll look at marking degraded and rebooting.

Reply:
Note that failure in `prepUpdateFromCluster()` will mark the node degraded on it's own. It doesn't, however, reboot. Will add that and some comments.